PyTorch Profiler Shape aggregation support

salexspb opened the pull request:
Summary:
This is useful when you would like to understand performance
bottlenecks of your model.  One can use the shape analysis in order to
fit model to a roofline model of their hardware.

Please note that this feature can potentially skew profiling
results. Also timing for not nested events will become wrong. One
should only use timing for the bottom most events when shape analysis
is used. Also for the case where people don't need shapes, profiling
should not be affected. As in this case we don't collect shapes, which
is the default behavior and this diff doesn't change it.

One of the next steps
could be, for example, choosing best candidates for quantization. In
the scope of this diff I am just adding optional shapes collection
into the Even class. After that in python there is minor functionality
for providing groupping by shapes.

In the output tables shapes are being truncated but in groupping full
shape string is used as a key.

Here is an example output:

test_profiler_shapes (test_autograd.TestAutograd) ...
```
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
unsigned short      2.30%            305.031us        2.30%            305.031us        305.031us        NaN              0.000us          0.000us          1                [[30, 20]]
addmm               69.40%           9.199ms          69.40%           9.199ms          9.199ms          NaN              0.000us          0.000us          1                [[30], [128, 20], [20, 30], [], []]
unsigned short      0.98%            129.326us        0.98%            129.326us        129.326us        NaN              0.000us          0.000us          1                [[40, 30]]
addmm               27.32%           3.621ms          27.32%           3.621ms          3.621ms          NaN              0.000us          0.000us          1                [[40], [128, 30], [30, 40], [], []]
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 13.255ms
CUDA time total: 0.000us

------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
unsigned short      2.30%            305.031us        2.30%            305.031us        305.031us        NaN              0.000us          0.000us          1                [[30, 20]]
addmm               69.40%           9.199ms          69.40%           9.199ms          9.199ms          NaN              0.000us          0.000us          1                [[30], [128, 20], [20, 30], [], []]
unsigned short      0.98%            129.326us        0.98%            129.326us        129.326us        NaN              0.000us          0.000us          1                [[40, 30]]
addmm               27.32%           3.621ms          27.32%           3.621ms          3.621ms          NaN              0.000us          0.000us          1                [[40], [128, 30], [30, 40], [], []]
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 13.255ms
CUDA time total: 0.000us
```

Also added this for older aggregation test:

```
test_profiler_aggregation_lstm (test_autograd.TestAutograd) ...
======================================================================================================================================================================================================
TEST
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
lstm                     0.69%            4.606ms          5.30%            35.507ms         35.507ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.67%            4.521ms          5.27%            35.340ms         35.340ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.66%            4.399ms          5.02%            33.638ms         33.638ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.65%            4.354ms          4.92%            32.958ms         32.958ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.65%            4.351ms          4.96%            33.241ms         33.241ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.65%            4.323ms          5.10%            34.163ms         34.163ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.304ms          4.92%            32.938ms         32.938ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.300ms          5.10%            34.172ms         34.172ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.292ms          5.05%            33.828ms         33.828ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.263ms          4.98%            33.357ms         33.357ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 670.120ms
CUDA time total: 0.000us

-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
sigmoid                  15.32%           102.647ms        15.32%           102.647ms        171.078us        NaN              0.000us          0.000us          600              [[3, 20]]
mul                      15.20%           101.854ms        15.20%           101.854ms        169.757us        NaN              0.000us          0.000us          600              [[3, 20], [3, 20]]
lstm                     12.74%           85.355ms         100.00%          670.120ms        33.506ms         NaN              0.000us          0.000us          20               [[5, 3, 10]]
addmm                    11.16%           74.808ms         11.16%           74.808ms         249.361us        NaN              0.000us          0.000us          300              [[80], [3, 20], [20, 80], [], []]
tanh                     9.89%            66.247ms         9.89%            66.247ms         165.617us        NaN              0.000us          0.000us          400              [[3, 20]]
split                    6.42%            43.019ms         6.42%            43.019ms         215.095us        NaN              0.000us          0.000us          200              [[3, 80]]
add                      5.67%            38.020ms         5.67%            38.020ms         190.101us        NaN              0.000us          0.000us          200              [[3, 80], [3, 80], []]
add                      4.81%            32.225ms         4.81%            32.225ms         161.124us        NaN              0.000us          0.000us          200              [[3, 20], [3, 20], []]
addmm                    3.79%            25.380ms         3.79%            25.380ms         253.796us        NaN              0.000us          0.000us          100              [[80], [3, 10], [10, 80], [], []]
unsigned short           3.72%            24.925ms         3.72%            24.925ms         83.083us         NaN              0.000us          0.000us          300              [[80, 20]]
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 670.120ms
CUDA time total: 0.000us

Total time based on python measurements:  691.366ms
CPU time measurement python side overhead: 3.17%
ok
```

Differential Revision: D15174987

ilia-cher commented:
@cpuhrsch Christian, I think this will be helpful for that use case of debugging shapes you mentioned

cpuhrsch commented:
Awesome!

stephenroller commented:
Just checking: does this make the fp16/fp32 distinction? It might be useful for making sure you're not accidentally slipping into fp32 ops at some point.

stephenroller commented:
@salexspb if it's easy to support an fp16/32 distinction it could be really useful for identifying:
1. If I'm converting a model to fp16, was I successful at finding anywhere I forgot to use .as_type(dtype) instead of .float(), or had to do some ops in fp32 temporarily but forgot to convert them back
2. Are the ops having the same profile ratios in fp16/fp32 (maybe an op is particularly inefficient in fp16 and I should avoid it).

Essentially grouping the shapes by dtype too would potentially be very beneficial

salexspb commented:
> @salexspb if it's easy to support an fp16/32 distinction it could be really useful for identifying:
> 
> 1. If I'm converting a model to fp16, was I successful at finding anywhere I forgot to use .as_type(dtype) instead of .float(), or had to do some ops in fp32 temporarily but forgot to convert them back
> 2. Are the ops having the same profile ratios in fp16/fp32 (maybe an op is particularly inefficient in fp16 and I should avoid it).
> 
> Essentially grouping the shapes by dtype too would potentially be very beneficial

I think this is a good idea, thanks for the suggestion! It should be one of the next follow up tasks. There are several things that need to happen. We need standardize the way we represent reduced precision ops. I don't know if there is a consensus here. In caffe2, for example, some ops do dynamic dispatch while others are special ops which have type name hardcoded into their name. 

If we want this to be more generic, we may just use a set of tensor data types as a key. Then dependency on operator naming is gone. But this is not being currently passed over to RecordFunction. It can be one more flag to do that along the lines @ilia-cher  implemented shape information passing to RecordFunction.

ilia-cher commented:
@salexspb we have not just shapes but whole tensors available in RecordFunction, so it should be trivial so save not only shape but also a dtype and pass it back to python?

cpuhrsch commented:
Another feedback I've just gotten is the ability to generate metrics like flops and memory bandwidth reliably as part of this. That means, a user gets a good flops and memory bandwidth estimation for each call.

salexspb commented:
@ilia-cher , you are right, my bad. Should be easy indeed. I'll have a look
@cpuhrsch , flops and memory reads / writes are hard. On CPU there are counters, but getting those right is not easy. And validation is tricky. In Caffe2 we used analytical formulas. This is something which would be useful to add to pytorch as well. In caffe2 it is a part of the operator schema. 

cc @dzhulgakov  @smessmer  about ways to integrate cost inference functions into pytorch. Was there any discussion about it already?

dzhulgakov commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -88,39 +95,57 @@ void popRange() {
+              continue;
+            }
+            const at::Tensor& tensor = input.toTensor();
+            if (tensor.defined()) {
```
you skip the input if tensor is undefined (do inputSizes.emplace_back())

dzhulgakov commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -48,12 +50,16 @@ void mark(std::string name, bool include_cuda /* = true */) {
-    if(sequence_nr >= 0) {
+    if (sequence_nr >= 0) {
       std::stringstream s;
       s << name.str() << msg << sequence_nr;
```
unrelated to this diff, but isn't msg here always the same constant? can we kill this argument?

salexspb replied:
not sure, I haven't tried using profiler on CUDA. @zdevito  might know

dzhulgakov commented on torch/csrc/autograd/init.cpp:
```diff
@@ -25,6 +26,15 @@ PyObject * THPAutograd_initExtension(PyObject *_unused)
 
   auto m = py::handle(autograd_module).cast<py::module>();
 
+  py::enum_<torch::autograd::profiler::ProfilerState>(m,"ProfilerState")
```
nit: remove namespace then

dzhulgakov reviewed (commented):
This is great!

> cc @dzhulgakov @smessmer about ways to integrate cost inference functions into pytorch. Was there any discussion about it already?

We should probably add an extra field in c10 for new ops. For existing ones, the easiest would be to put it next to operators in native_functions.yaml (and then we can codegen the c10 bits too)

soumith commented on test/test_autograd.py:
```diff
@@ -2467,8 +2491,10 @@ def test_profiler_aggregation_lstm(self):
 
-        print(prof.table(sort_by="self_cpu_time_total", row_limit=10))
-        print(prof.key_averages().table(sort_by="self_cpu_time_total", row_limit=10))
+        print(prof.table(
```
actually I am not sure why we were printing to stdout before, it's not useful in the unit test. can you change it, just call `__repr__` on it for the sanity check.

salexspb replied:
@soumith , is there a specific issue with printing to stdout? Normally a python test runner would capture std out. But when you run a test manually output is useful. Is this is not a case for some of the OSS CIs? 
 
No strong feelings about this though :)

soumith replied:
the default testrunner for python doesn't capture stdout. But the default one in `buck` eats it properly.
Anyways, this is a nit and it wasn't introduced in your diff.

soumith commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -88,39 +95,59 @@ void popRange() {
   AT_ASSERT(new_state != ProfilerState::Disabled);
   if (new_state == ProfilerState::NVTX && !cuda_stubs->enabled())
-    throw std::runtime_error("Can't use NVTX profiler - PyTorch was compiled without CUDA");
+    throw std::runtime_error(
```
nit: this diff is sprinkled with formatting changes that are unnecessary. pytorch's column limit is 120 chars, so all of the 80-limit changes are unncessary

salexspb replied:
@soumith , I agree that it makes things harder for review. This is an issue with tooling I see here.. I apply standard Facebook code formatting to a block of code I change. But it can touch other parts. I didn't reformat the whole file.

for me it is hard to spend time manually undoing changes which are being done automatically by our Atom integration. Do we have a good wiki page which describes how one can integrate pytorch code formatting into Atom?

salexspb replied:
btw, the code looked really bad below. Like doing "for()" instead of "for (...)". This is not related to 120 characters limit.

soumith replied:
I thought we made sure the flake / linter in fbcode is now properly specified to respect some pytorch formatting. cc: @ezyang if you have any suggestions here.

salexspb replied:
Well, my use-case is the following. I type code, it causes linter warning or just looks ugly. I select this code, type cmd + c and it formats the code. Sometimes it can touch more stuff for some reason. Btw, things that it touched are not due to 120 characters mainly but dew to missing whitespaces. Which are really annoying and make the code hard to read. 

IMO if this is just a few places which make code nicers - it is fine. If this is a really large file being fully rewritten - thats different.

ezyang replied:
I've primarily looked at the Python lint situation, and I am not familiar with what is going on with the C++.

But there's a very simple fix for this problem; just do the reformatting as a commit beforehand, and then stack the actual change on top.

soumith commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -88,39 +95,59 @@ void popRange() {
-    popRange();
-  });
+  pushCallback(
+      [config](const RecordFunction& fn) {
```
profiling with shape recording enabled will have much more overhead than without. is it worth leaving a note in the python docstring saying so?

salexspb replied:
Good point, I will add it

zdevito reviewed (approved):
Code looks good to me.

> Also timing for not nested events will become wrong.
I don't understand why this is true. If it is, can this be added to the documentation with an explaination.

salexspb commented:
Chatted with @zdevito , the docs are probably fine as they are. The answer to why I warn more about top functions is that if function  calls other functions, the overhead of shape collection inside internal functions can be added to self cpu time of the parent function. While the bottom most function's self cpu time stays unaffected by shape vector copy.

salexspb committed:
PyTorch Profiler Shape aggregation support (#20035)

Summary:
This is useful when you would like to understand performance
bottlenecks of your model.  One can use the shape analysis in order to
fit model to a roofline model of their hardware.

Please note that this feature can potentially skew profiling
results. Also timing for not nested events will become wrong. One
should only use timing for the bottom most events when shape analysis
is used. Also for the case where people don't need shapes, profiling
should not be affected. As in this case we don't collect shapes, which
is the default behavior and this diff doesn't change it.

One of the next steps
could be, for example, choosing best candidates for quantization. In
the scope of this diff I am just adding optional shapes collection
into the Even class. After that in python there is minor functionality
for providing groupping by shapes.

In the output tables shapes are being truncated but in groupping full
shape string is used as a key.

Here is an example output:

test_profiler_shapes (test_autograd.TestAutograd) ...
```
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
unsigned short      2.30%            305.031us        2.30%            305.031us        305.031us        NaN              0.000us          0.000us          1                [[30, 20]]
addmm               69.40%           9.199ms          69.40%           9.199ms          9.199ms          NaN              0.000us          0.000us          1                [[30], [128, 20], [20, 30], [], []]
unsigned short      0.98%            129.326us        0.98%            129.326us        129.326us        NaN              0.000us          0.000us          1                [[40, 30]]
addmm               27.32%           3.621ms          27.32%           3.621ms          3.621ms          NaN              0.000us          0.000us          1                [[40], [128, 30], [30, 40], [], []]
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 13.255ms
CUDA time total: 0.000us

------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
unsigned short      2.30%            305.031us        2.30%            305.031us        305.031us        NaN              0.000us          0.000us          1                [[30, 20]]
addmm               69.40%           9.199ms          69.40%           9.199ms          9.199ms          NaN              0.000us          0.000us          1                [[30], [128, 20], [20, 30], [], []]
unsigned short      0.98%            129.326us        0.98%            129.326us        129.326us        NaN              0.000us          0.000us          1                [[40, 30]]
addmm               27.32%           3.621ms          27.32%           3.621ms          3.621ms          NaN              0.000us          0.000us          1                [[40], [128, 30], [30, 40], [], []]
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 13.255ms
CUDA time total: 0.000us
```

Also added this for older aggregation test:

```
test_profiler_aggregation_lstm (test_autograd.TestAutograd) ...
======================================================================================================================================================================================================
TEST
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
lstm                     0.69%            4.606ms          5.30%            35.507ms         35.507ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.67%            4.521ms          5.27%            35.340ms         35.340ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.66%            4.399ms          5.02%            33.638ms         33.638ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.65%            4.354ms          4.92%            32.958ms         32.958ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.65%            4.351ms          4.96%            33.241ms         33.241ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.65%            4.323ms          5.10%            34.163ms         34.163ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.304ms          4.92%            32.938ms         32.938ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.300ms          5.10%            34.172ms         34.172ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.292ms          5.05%            33.828ms         33.828ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
lstm                     0.64%            4.263ms          4.98%            33.357ms         33.357ms         NaN              0.000us          0.000us          1                [[5, 3, 10]]
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 670.120ms
CUDA time total: 0.000us

-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  Input Shapes
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
sigmoid                  15.32%           102.647ms        15.32%           102.647ms        171.078us        NaN              0.000us          0.000us          600              [[3, 20]]
mul                      15.20%           101.854ms        15.20%           101.854ms        169.757us        NaN              0.000us          0.000us          600              [[3, 20], [3, 20]]
lstm                     12.74%           85.355ms         100.00%          670.120ms        33.506ms         NaN              0.000us          0.000us          20               [[5, 3, 10]]
addmm                    11.16%           74.808ms         11.16%           74.808ms         249.361us        NaN              0.000us          0.000us          300              [[80], [3, 20], [20, 80], [], []]
tanh                     9.89%            66.247ms         9.89%            66.247ms         165.617us        NaN              0.000us          0.000us          400              [[3, 20]]
split                    6.42%            43.019ms         6.42%            43.019ms         215.095us        NaN              0.000us          0.000us          200              [[3, 80]]
add                      5.67%            38.020ms         5.67%            38.020ms         190.101us        NaN              0.000us          0.000us          200              [[3, 80], [3, 80], []]
add                      4.81%            32.225ms         4.81%            32.225ms         161.124us        NaN              0.000us          0.000us          200              [[3, 20], [3, 20], []]
addmm                    3.79%            25.380ms         3.79%            25.380ms         253.796us        NaN              0.000us          0.000us          100              [[80], [3, 10], [10, 80], [], []]
unsigned short           3.72%            24.925ms         3.72%            24.925ms         83.083us         NaN              0.000us          0.000us          300              [[80, 20]]
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 670.120ms
CUDA time total: 0.000us

Total time based on python measurements:  691.366ms
CPU time measurement python side overhead: 3.17%
ok
```
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20035

Differential Revision: D15174987

fbshipit-source-id: 3995f5e7fa891b755207e3024664d94204442896
